client/asset/dcr/dcr.go

```diff
@@ -4908,3 +4932,196 @@ func float64PtrStr(v *float64) string {
 	}
 	return strconv.FormatFloat(*v, 'f', 8, 64)
 }
+
+// ConfirmRedemption returns how many confirmations a redemption has. Normally
+// this is very straitforward. However there are two situations that have come
+// up that this also handles. One is when the wallet can not find the redemption
+// transaction. This is most likely because the fee was set too low and the tx
+// was removed from the mempool. In the case where it is not found, this will
+// send a new fee tx using the provided fee suggestion. The second situation
+// this watches for is a transaction that we can find but has been sitting in
+// the mempool for a long time. This has been observed with the wallet in SPV
+// mode and the transaction inputs having been spent by another transaction. The
+// wallet will not pick up on this so we tell it to abandon the original
+// transaction and, again, send a new one using the provided feeSuggestion.
+func (dcr *ExchangeWallet) ConfirmRedemption(coinID dex.Bytes, redemption *asset.Redemption, feeSuggestion uint64) (*asset.ConfirmRedemptionStatus, error) {
+	txHash, _, err := decodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+
+	var secretHash [32]byte
+	copy(secretHash[:], redemption.Spends.SecretHash)
+	dcr.mempoolRedeemsMtx.RLock()
+	mRedeem, have := dcr.mempoolRedeems[secretHash]
+	dcr.mempoolRedeemsMtx.RUnlock()
+
+	var deleteMempoolRedeem bool
+	defer func() {
+		if deleteMempoolRedeem {
+			dcr.mempoolRedeemsMtx.Lock()
+			delete(dcr.mempoolRedeems, secretHash)
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+	}()
+
+	tx, err := dcr.wallet.GetTransaction(dcr.ctx, txHash)
+	if err != nil && !errors.Is(err, asset.CoinNotFoundError) {
+		return nil, fmt.Errorf("problem searching for redemption transaction %s: %w", txHash, err)
+	}
+	if err == nil {
+		var abandonTx bool
+		if have && mRedeem.txHash == *txHash {
+			if tx.Confirmations == 0 && time.Now().After(mRedeem.firstSeen.Add(maxRedeemMempoolAge)) {
+				// Transaction has been sitting in the mempool
+				// for a long time now. Attempt to abandon.
+				abandonTx = true
+				if err := dcr.wallet.AbandonTransaction(dcr.ctx, txHash); err != nil {
+					dcr.log.Warnf("Unable to abandon redemption mempool transaction %v: %v.", txHash, err)
+				} else {
+					dcr.log.Infof("Abandoned redemption mempool transaction: %v.", txHash)
+					deleteMempoolRedeem = true
+				}
+			}
+		} else {
```

Comment:
`else if !have`?

Reply:
If `have` is true here, I think it is an error. Core has given this method a different redeem for this secret hash. It is this method's job to assign a new transaction so something is off. Perhaps the method was run concurrently for the same redeem and this is the result of a race if hit. Adding an error return here so we do not erase what the first call to this method must have done.